### PR TITLE
EDoc spec cleanup

### PR DESCRIPTION
**Pull request by erszcz**
This extracts spec/type changes from #2803 for easier review. No functional changes to EDoc are made in this PR.

### CODE REVIEW

**Comment by erszcz** [comment]
FYI, I'm working on the broken links.

**Comment by erszcz** [comment]
Apparently, it's not possible to take just the spec/type changes. I tried to make them as nonintrusive as possible.

File: lib/xmerl/src/xmerl_scan.erl [hunk 1]
```diff
@@ -140,11 +140,14 @@
 %% helper functions. To xmerl_lib ??
 -export([accumulate_whitespace/4]).
 
+-export_type([xmlElement/0]).
+
 %-define(debug, 1).
 -include("xmerl.hrl").		% record def, macros
 -include("xmerl_internal.hrl").
 -include_lib("kernel/include/file.hrl").
```
Comment (garazdawi) [inline lib/xmerl/src/xmerl_scan.erl:149]: @lthor do you have any feedback about the addition of this type?

**Review by garazdawi — COMMENTED** [review]
I think this looks good.

Do we need tests for the app inference logic, or is it tested by building docs for Erlang/OTP?

**Comment by erszcz** [comment]
> Do we need tests for the app inference logic, or is it tested by building docs for Erlang/OTP?

For OTP apps, if the links were wrong CI would fail. That's why https://github.com/erlang/otp/pull/2914/commits/de12adfa521e938694933bdbe94b576e1966e556 originally failed with "Broken links" and "References to missing anchors".

In general, this inference is used when from one app (the app currently processed by EDoc) we refer to a type defined in a module outside the app. If the external module is preloaded or not found under `external_app-1.2.3/ebin/external_mod.beam`, the inference will return `no_app`, therefore the generated link will lead to the currently processed app. This is wrong, but also quite rare and I'm not sure how to address it :| For example, this would occur if in `edoc.erl` we had `-type t() :: zlib:zflush().`, since `zlib` is preloaded.

Outside OTP, it should work fine for as long as we refer to modules in properly structured apps, i.e. following `app-1.2.3/ebin/mod.beam.`

**Comment by erszcz** [comment]
> This is wrong, but also quite rare and I'm not sure how to address it :| 

BTW, this is also the status quo if the inference isn't merged in.

File: lib/edoc/src/edoc.erl [hunk 2]
```diff
@@ -60,23 +60,71 @@
...
+-type edoc_module() :: xmerl_scan:xmlElement().
+%% The EDoc documentation data for a module,
+%% expressed as an XML document in {@link //xmerl. XMerL} format. See
+%% the file <a href="edoc.dtd">`edoc.dtd'</a> for details.
+
+-type ordset(T) :: ordsets:ordset(T).
+-type function_name() :: {atom(), integer()}.
+-type filename() :: file:filename().
+-type proplist() :: proplists:proplist().
+-type comment2() :: { Line :: integer(),
```
Comment (garazdawi) [inline lib/edoc/src/edoc.erl:114]: Can we give this a better name than `comment2`?

Comment (erszcz) [reply lib/edoc/src/edoc.erl:114]: Definitely, just a second...

Comment (erszcz) [reply lib/edoc/src/edoc.erl]: I've pushed a rename. I've also found out that this type is actually an alias of `erl_comment_scan:comment()`, which is exported already. The data instances are read by `erl_comment_scan:read_comments()`. Should I mark it as `-type comment() :: erl_comment_scan:comment()`?

Comment (garazdawi) [reply lib/edoc/src/edoc.erl]: yes, I think that would be better.

**Comment by erszcz** [comment]
> If the external module is preloaded or not found under `external_app-1.2.3/ebin/external_mod.beam`, the inference will return `no_app`

It just occurred to me that some preloaded modules end up having a corresponding `.beam` file under the `erts` app:

```
$ find ~/work/erlang/build.otp -name zlib.beam
/Users/erszcz/work/erlang/build.otp/lib/erlang/lib/erts-11.1.3/ebin/zlib.beam
```

 If that holds for all preloaded modules, then the inference could be improved.

**Comment by garazdawi** [comment]
They are there for tools (such as dialyzer) to be able to get the debug_info when analyzing a released system. I don't think that they are guaranteed to be there, but it may very well be the case that in the scenarios that count they are.

**Comment by erszcz** [comment]
I've added a test following the above assumptions, with the small difference that when tests are run from the source tree, `code_lib:dir(erts)` doesn't have a direct `ebin` subdir. This change should make it possible to properly link to types like `erlang:spawn_opt_type/0` in edoc comments.

**Comment by erszcz** [comment]
Hmm, the last build failed, but it seems to be an environment issue - `Error: read ECONNRESET` - not actually connected with the new changes 🤔

**Comment by garazdawi** [comment]
I restarted the jobs

**Comment by garazdawi** [comment]
I'm happy with these changes now. Would you mind reworking the commit history so that there are only functional working commits? When you are done I will merge this.

**Comment by garazdawi** [comment]
I'll put it in testing once more just to make sure nothing has been broken by the rebase and then merge tomorrow.

**Comment by erszcz** [comment]
Sounds good, thanks for the review!

**Pull request merged**